swarm/pss/pss.go

```diff
@@ -241,91 +317,283 @@ func (self *Pss) checkFwdCache(addr []byte, digest pssDigest) bool {
+		KeySym:   symkey,
+		Topic:    topic,
+		WorkTime: defaultWhisperWorkTime,
+		PoW:      defaultWhisperPoW,
+		Payload:  msg,
+		Padding:  []byte("1234567890abcdef"),
+	}
+	// set up outgoing message container, which does encryption and envelope wrapping
+	woutmsg, err := whisper.NewSentMessage(wparams)
+	if err != nil {
+		return fmt.Errorf("failed to generate whisper message encapsulation: %v", err)
+	}
+	// performs encryption and PoW
+	// after this the message is ready for sending
+	envelope, err := woutmsg.Wrap(wparams)
+	if err != nil {
+		return fmt.Errorf("failed to perform whisper encryption: %v", err)
+	}
+	log.Trace("pssmsg whisper done", "env", envelope, "wparams payloda", wparams.Payload)
+	// prepare for devp2p transport
 	pssmsg := &PssMsg{
 		To:      to,
-		Payload: pssenv,
+		Payload: envelope,
 	}
 	return self.Forward(pssmsg)
 }
 
-// Forwards a pss message to the peer(s) closest to the to address
+// Forwards a pss message to the peer(s) closest to the to recipient address in the PssMsg struct
+//
+// The To-field in the PssMsg struct will be plaintext, unencrypted, and will be truncated to the magnitude set by SetRecipientAddressLength (PssParams.RecipientAddressLength by default).
+// If the recipientAddressLength value is 0, no bytes will be disclosed, and message will be forwarded to ALL devp2p peers
 //
-// Handlers that are merely passing on the PssMsg to its final recipient should call this directly
+// Handlers that are merely passing on the PssMsg to its final recipient might call this directly
 func (self *Pss) Forward(msg *PssMsg) error {
-
-	if self.isSelfRecipient(msg) {
-		return errorForwardToSelf
-	}
-
-	// cache it
+	to := msg.To
+	zeros := make([]byte, len(to)-int(self.recipientAddressLength))
+	// truncate part of the pssmsg recipient address
+	// this way noone can know for certain exactly who it's for
+	// the kademlia address will be same length
+	// but equivalent to truncated part will be overwritten by zeros
+	copy(to[self.recipientAddressLength:len(to)], zeros)
+	log.Trace("truncated recipient", "original", msg.To, "truncated", to)
+	msg.To = msg.To[:self.recipientAddressLength]
+
+	// cache the message
 	digest, err := self.storeMsg(msg)
 	if err != nil {
 		log.Warn(fmt.Sprintf("could not store message %v to cache: %v", msg, err))
 	}
 
-	// flood guard
+	// flood guard:
+	// don't allow identical messages we saw shortly before
 	if self.checkFwdCache(nil, digest) {
 		log.Trace(fmt.Sprintf("pss relay block-cache match: FROM %x TO %x", common.ByteLabel(self.Overlay.BaseAddr()), common.ByteLabel(msg.To)))
 		return nil
 	}
 
-	// TODO:check integrity of message
-	sent := 0
-
 	// send with kademlia
 	// find the closest peer to the recipient and attempt to send
-	self.Overlay.EachConn(msg.To, 256, func(op network.OverlayConn, po int, isproxbin bool) bool {
+	sent := 0
+
+	self.Overlay.EachConn(to, 256, func(op network.OverlayConn, po int, isproxbin bool) bool {
```

Comment:
that is the whole point of partial addressing yes. 
closer than the lumunosity limit expressed by the visible length of the address, you have no idea where to send the message, so you send it to everyone 

i.e., pss becomes shh (whisper), deterministic routing is replaced with broadcast/flooding/gossip, 

when there is too much light on any one potential recipient (the distinguishability order (log scale) is too low, i.e., the cost of cracking anonymity) the message must be sent in the dark

Ah and another thought, of course if you are the recipient, partial addressed envelopes need to be forwarded as well as processed.